db/db_test.go

```diff
@@ -564,7 +574,65 @@ var _ = Describe("Database operation test", func() {
 						return nil
 					}
 					err := secondTx.Exec(ctx, "update todos set description = 'other_description' where id = 'second'")
-					Expect(db.IsDeadlock(err)).To(BeTrue())
+					Expect(db.IsKnownDatabaseError(err)).To(BeTrue())
+					return err
+				})
+				Expect(err).To(Succeed())
+				return nil
+			})).To(Succeed())
+		})
+	})
+
+	Context("MySQL transaction retries", func() {
+		var (
+			firstConn  db.DB
+			secondConn db.DB
+
+			connOpts = options.Options{
+				RetryTxCount:    5,
+				RetryTxInterval: 100 * time.Millisecond,
+			}
+		)
+		JustBeforeEach(func() {
+			firstConn, err = dbutil.ConnectDB(mysqlType, mysqlConn, db.DefaultMaxOpenConn, connOpts)
+			Expect(err).ToNot(HaveOccurred())
+			secondConn, err = dbutil.ConnectDB(mysqlType, mysqlConn, db.DefaultMaxOpenConn, connOpts)
+			Expect(err).ToNot(HaveOccurred())
+
+			for _, s := range manager.OrderedSchemas() {
+				Expect(firstConn.RegisterTable(s, false, true)).To(Succeed())
+			}
+		})
+
+		AfterEach(func() {
+			firstConn.Close()
+			secondConn.Close()
+			schema.ClearManager()
+		})
+
+		BeforeEach(func() {
+			initData()
+		})
+
+		It("should retry a few times after a deadlock", func() {
```

Comment:
This test depends on other tests:
```
$ go test -count=1 -ginkgo.focus "MySQL transaction retries"
[...]
• Failure [0.352 seconds]
Database operation test
/home/tomasz/code/src/github.com/cloudwan/gohan/db/db_test.go:38
  MySQL transaction retries
  /home/tomasz/code/src/github.com/cloudwan/gohan/db/db_test.go:586
    should retry a few times after a deadlock [It]
    /home/tomasz/code/src/github.com/cloudwan/gohan/db/db_test.go:617

    Expected
        <uint64>: 0
    to equal
        <uint64>: 2

    /home/tomasz/code/src/github.com/cloudwan/gohan/db/db_test.go:623
```
Note that `initData()` does not actually create resources in database (it only initializes suite variables).

Reply:
Other tests can be fixed by dropping tables in `AfterEach` (they are created in `BeforeEach`), e.g.

```
schemas := manager.OrderedSchemas()
for i := len(schemas) - 1; i >= 0; i-- {
	Expect(dataStore.DropTable(schemas[i])).To(Succeed())
}
```